docs/SPEC.md

```diff
@@ -1364,6 +1371,191 @@ A query specification defines what data and operations to perform.
 The execution model reserves the right to perform those operations as efficiently as possible.
 The execution model may rewrite the query in anyway it sees fit while maintaining correctness.
 
+## Modules
+
+>NOTE: Modules are not fully implemented yet, follow https://github.com/influxdata/flux/issues/4296 for details.
+
+A module is a collection of packages that can be imported.
+A module has a module path, version and a collection of packages with their source code.
+
+### Module path
+
+The module path is the import path of the top level package within the module.
+Additionally major versions of a module of two or greater must add a final element to the path of the form `v#` where `#` is the major version number.
+For modules at version zero or one the path must not contain the major version as it is not necessary.
+A change from `v0` to `v1` may include a breaking change but once `v1` is published any future breaking changes will be a new major version.
+
+Example
+
+    foo/bar    // module path of foo/bar for version zero or one
+    foo/bar/v2 // module path of foo/bar for major version two
+
+### Module versions
+
+All modules are versioned using a [semantic version number](https://semver.org/) prefixed with a `v`, i.e. `vMAJOR.MINOR.PATCH`.
+Once a module version has been published it cannot be modified.
+A new version of the module must be published.
+
+### Importing modules
+
+Flux modules are imported using an import declaration.
+The import path may contain specifiers about which versions of a module should be imported.
+
+When no version information is provided the latest version of the module is used.
+A _minimum_ version may be specified using `@` followed by a version number.
+An import may also specify the version `pre` which is the most recent pre-release version of the module.
+
+Examples
+
+    import "foo/module"             // imports the latest version 0.x or 1.x version of the module
```

Comment:
This will be ambiguous with standard library modules as it is currently (did you mean the user module "array" or the standard library "array"?). For the existing standard library we can just reserve the names for all existing packages, but that won't work if we want to add a new packages so maybe we also want to add and reserve a `std` prefix for the standard library?

Reply:
I've seen things in the JS space where importers use like `~/...` to say, "relative to the project's `node_modules` dir."

It could be useful to have something as a prefix to indicate the importer to use (registry vs stdlib).